### PR TITLE
Reducing allocations in value providers

### PR DESCRIPTION
**Pull request by rynowak**
- Don't go async in formdata providers unless we need to
- Remove unnecessary defensive copy in CompositeValueProvider

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs [hunk 1]
```diff
@@ -30,8 +30,8 @@ public CompositeValueProvider()
         /// </summary>
         /// <param name="valueProviders">The sequence of <see cref="IValueProvider"/> to add to this instance of
         /// <see cref="CompositeValueProvider"/>.</param>
-        public CompositeValueProvider(IEnumerable<IValueProvider> valueProviders)
-            : base(valueProviders.ToList())
+        protected CompositeValueProvider(IList<IValueProvider> valueProviders)
+            : base(valueProviders)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs]: Turning this into `IList<T>` avoids the copy - moreover we never even call this constructor outside of the class and some tests.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs]: The important note here is that when using uber-binding attributes like `[FromForm]` we'll instantiate a new CVP for each attribute. So you can allocate N+1 copies of this per request, and N+1 calls to trivial ToList()

**Comment by rynowak** [comment]
Working on getting some perf numbers for this change.

**Comment by rynowak** [comment]
**Before:**
image

**After**
image


This ended up being only worth about 400kb across 3000 requests, so not super impactful. Still seems worth taking given how simple the code changes are.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs]: Suggest `IReadOnlyList` and sub-classing `ReadOnlyCollection` to express intent a bit better. That is, unless I'm forgetting something that updates the `CompositeValueProvider`.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeValueProvider.cs]: We have product code that updates the `CompositeValueProvider`. I don't see much reason to change this detail or the usage pattern in that part of the code

**Comment by rynowak** [comment]
@dougbu - any other feedback?

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**